reference/decoder.go

```diff
@@ -0,0 +1,287 @@
+}
+
+func (v decoder) parseSchema(scheme, refFull string) (ByteDecodeFunc, error) {
+	parts := strings.Split(scheme, "+")
+	switch len(parts) {
+	case 1:
+		if parts[0] == SchemaV1 {
+			return v.defaultDecoder, nil
+		}
+	case 2:
+		switch {
+		case parts[0] == SchemaV1:
+			parts[0] = parts[1]
+		case parts[1] == SchemaV1:
+			//
+		default:
+			return nil, fmt.Errorf("unsupported scheme: scheme=%s, ref=%s", scheme, refFull)
+		}
+	default:
+		return nil, fmt.Errorf("invalid scheme: scheme=%s", scheme)
+	}
+	decoder := v.byteDecoderFactory.GetByteDecoder(parts[0])
+	if decoder == nil {
+		return nil, fmt.Errorf("unknown decoder: scheme=%s, decoder=%s, ref=%s", scheme, parts[0], refFull)
+	}
+	return decoder, nil
+}
+
+func (v decoder) parseAuthority(ref string) (authority string, remaining string) {
+	if len(ref) < 3 || ref[:2] != "//" {
+		return "", ref
+	}
+	ref = ref[2:]
+
+	pos := strings.IndexRune(ref, '/')
+	if pos < 0 {
+		return ref, ""
+	}
+
+	return ref[:pos], ref[pos+1:]
+}
+
+func (v decoder) parseReference(refFull string, byteDecoder ByteDecodeFunc) (Global, error) {
+	_, ref := v.parseAuthority(refFull)
+	if len(ref) == 0 {
+		return Global{}, fmt.Errorf("empty reference body: ref=%s", refFull)
+	}
+
+	parityPos := strings.IndexRune(ref, '/')
+	var parity []byte
+	switch {
+	case parityPos == 0:
+		return Global{}, fmt.Errorf("empty reference body: ref=%s", refFull)
+	case parityPos > 0:
+		buf := bytes.NewBuffer(make([]byte, 0, pulseAndScopeSize))
+		_, err := byteDecoder(ref[parityPos+1:], buf)
+		if err != nil {
+			return Global{}, fmt.Errorf("unable to decode parity: ref=%s, err=%v", refFull, err)
+		}
+		ref = ref[:parityPos]
+		if v.options&IgnoreParity == 0 {
+			parity = buf.Bytes()
+		}
+	}
+
+	var result Global
+	err := v.parseAddress(ref, byteDecoder, &result)
+
+	if err == nil && parity != nil {
+		err = result.CheckParity(parity)
+	}
+	if err == nil {
+		return result, nil
+	}
+
+	return result, fmt.Errorf("invalid reference, %s: ref=%s", err.Error(), refFull)
+}
+
+func (v decoder) parseAddress(ref string, byteDecoder ByteDecodeFunc, result *Global) error {
+
+	domainPos := strings.IndexRune(ref, '.')
+	switch {
+	case domainPos == 0:
+		return errors.New("empty reference body")
+	case domainPos > 0:
+		domainName := ref[domainPos+1:]
+		localAddrName := ref[:domainPos]
+		switch domainName {
+		case "":
+			return errors.New("empty domain name")
+		case RecordDomainName:
+			if v.options&AllowRecords == 0 {
+				return errors.New("record reference is not allowed")
+			}
+			return v.parseBinaryAddress(localAddrName, byteDecoder, &result.addressLocal)
+		case LegacyDomainName:
+			return errors.New("legacy domain name")
+		default:
+			resolveBase := &Global{}
+
+			err := v.parseBinaryAddress(domainName, byteDecoder, &resolveBase.addressLocal)
+			switch {
+			case err == nil:
+				if !resolveBase.tryConvertToSelf() {
+					return errors.New("invalid self reference")
```

Comment:
msg should say "invalid reference base", as base should always convertible to self

Reply:
fixed